### PR TITLE
WIP - Working multi-snap cluster

### PR DESCRIPTION
**Pull request by lazypower**
DO NOT MERGE!

This is an open thread to track the progress and discuss the approach to the refactoring of the etcd layer to migrate from apt delivery => snaps.

This has implications on:

- delivery mechanism
- backup/restore
- durable storage migration(s)

Initial deployment + upgrade code. Functional X node cluster. Does not survive leader termination.

### CODE REVIEW

File: templates/defaults [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # This file is rendered by Juju, manual edits will not be persisted
-ETCD_DATA_DIR=/var/lib/etcd/default
+ETCD_DATA_DIR={{ etcd_data_dir }}/{{ unit_name }}.etcd
```
Comment (lazypower) [inline templates/defaults:2]: This now matches what the snap is doing. I'm not 100% positive we want to keep this nomenclature but I'm for keeping it in lockstep with the snap so it stays consistent.

File: reactive/etcd.py [hunk 2]
```diff
@@ -61,21 +60,25 @@ def check_cluster_health():
     except Exception:
         peers = 0
 
-    status_message = "{0} with {1} known peers.".format(unit_health, peers)
+    if peers > 1:
+        subject = "peers"
+    else:
+        subject = "peer"
+
+    status_message = "{0} with {1} known {2}.".format(unit_health, peers,
```
Comment (marcoceppi) [inline reactive/etcd.py]: Could we do this instead?

```python
status_message = '{0} with {1} known peer{2}'.format(unit_health, peers, 's' if peers != 1 else '')
```

Comment (lazypower) [reply reactive/etcd.py]: We could, I don't know what it buys us other than collapsing four lines into one, but at the sake of readability. I don't feel strongly either way however.

File: actions/snap-upgrade [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# Handle the snap ingestion routine
+function ingest_snap {
+  # This determines if we have a resource provided etcd ingest snap.
+  if [ "${USE_RESOURCE}" == "False" ]; then
+    # If no manual snap specified, assume snapstore connectivity
+    snap install etcd --channel=ingest/stable --classic
+  else
+    # Install from the resource location, and check for correct snap.
+    MANUAL_SNAP=$(resource-get etcd)
+    snap install ${ETCD_MANUAL_SNAP} --classic
```
Comment (adam-stokes) [inline actions/snap-upgrade]: If you are installing a local snap you need to pass `--dangerous` as well

File: actions/snap-upgrade [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+if [ ! $(charms.reactive is_state etcd.snap.ingest) ]; then
+  ingest_snap
+fi
+
+if [ "${USE_RESOURCE}" == "False" ]; then
+  CONFIGURED_CHANNEL=$(config-get channel)
+  snap refresh etcd --channel=${CONFIGURED_CHANNEL} --classic
+else
+  MANUAL_SNAP=$(resource-get etcd)
+  snap refresh ${MANUAL_SNAP}
```
Comment (adam-stokes) [inline actions/snap-upgrade]: Also i dont think snap refresh works for locally installed snaps

Comment (marcoceppi) [reply actions/snap-upgrade]: You'll need to just snap install /again/

**Comment by lazypower** [comment]
Closing for a formal PR

**Pull request closed without merging**